src/run-graph.ts

```diff
@@ -305,12 +308,30 @@ export class RunGraph {
     return pkgsInError.length > 0
   }
 
-  expandGlobs(globs: string[]) {
-    return this.pkgJsons.map(p => p.name).filter(name => globs.some(glob => minimatch(name, glob)))
+  async expandGlobs(globs: string[]) {
+    let pkgs = this.pkgJsons
+      .map(p => p.name)
+      .filter(name => globs.some(glob => minimatch(name, glob)))
+
+    // if changedSince is defined, filter the packages to contain only changed packages (according to git)
+    if (this.opts.changedSince) {
+      return getChangedFilesForRoots([this.opts.workspacePath], {
+        changedSince: this.opts.changedSince
+      }).then(data => {
+        if (!data.repos || (data.repos.git.size === 0 && data.repos.hg.size === 0)) {
+          throw new Error(
+            "The workspace is not a git/hg repo and it cannot work with 'changedSince'"
+          )
+        }
+
+        return filterChangedPackages([...data.changedFiles], this.pkgPaths, this.opts.workspacePath)
```

Comment:
Is the return type Promise<PkgPaths[]> ?

Reply:
No. The `expandGlobs` function returns `Promise<string[]>` (in master) so the new function `filterChangedPackages` has the same signature.

Reply:
ah since it `.map`s `p => p.name` first. got it.